src/main/scala/org/graphframes/examples/BeliefPropagation.scala

```diff
@@ -0,0 +1,259 @@
+              val newBelief = math.exp(-log1pExp(-x))
+              VertexAttr(vAttr.a, newBelief, color)
+            } else {
+              vAttr
+            }
+        }
+      }
+    }
+
+    // Convert back to GraphFrame with a new column "belief" for vertices DataFrame.
+    val gxFinal: Graph[Double, Unit] = gx.mapVertices((_, attr) => attr.belief).mapEdges(_ => ())
+    GraphFrame.fromGraphX(colorG, gxFinal, vertexNames = Seq("belief"))
+  }
+
+  case class VertexAttr(a: Double, belief: Double, color: Int)
+
+  case class EdgeAttr(b: Double)
+
+  /**
+   * Run Belief Propagation.
+   *
+   * This implementation of BP shows how to use GraphFrame's aggregateMessages method.
+   *  - Color GraphFrame vertices for BP scheduling.
+   *  - Run BP using GraphFrame's aggregateMessages API.
+   *  - Augment the original GraphFrame with the BP results (vertex beliefs).
+   *
+   * @param g  Graphical model created by [[org.graphframes.examples.Graphs.gridIsingModel()]]
+   * @param numIter  Number of iterations of BP to run.  One iteration includes updating each
+   *                 vertex's belief once.
+   * @return  Same graphical model, but with [[GraphFrame.vertices]] augmented with a new column
+   *          "belief" containing P(x,,i,, = +1), the marginal probability of vertex i taking
+   *          value +1 instead of -1.
+   */
+  def runBPwithGraphFrames(g: GraphFrame, numIter: Int): GraphFrame = {
+    // Choose colors for vertices for BP scheduling.
+    val colorG = colorGraph(g)
+    val numColors: Int = colorG.vertices.select("color").distinct.count().toInt
+
+    // TODO: Handle vertices without any edges.
+
+    // Initialize vertex beliefs at 0.0.
+    var gx = GraphFrame(colorG.vertices.withColumn("belief", lit(0.0)), colorG.edges)
+
+    // Run BP for numIter iterations.
+    for (iter <- Range(0, numIter)) {
+      // For each color, have that color receive messages from neighbors.
+      for (color <- Range(0, numColors)) {
+        // Send messages to vertices of the current color.
+        val AM = AggregateMessages
+        // Can send to source or destination since edges are treated as undirected.
+        val msgForSrc: Column = when(AM.src("color") === color, AM.edge("b") * AM.dst("belief"))
+        val msgForDst: Column = when(AM.dst("color") === color, AM.edge("b") * AM.src("belief"))
+        val logistic = udf { (x: Double) => math.exp(-log1pExp(-x)) }
+        val aggregates = gx.aggregateMessages
+          .sendToSrc(msgForSrc)
+          .sendToDst(msgForDst)
+          .agg(sum(AM.msg).as("aggMess"))
+        val v = gx.vertices
+        // Receive messages, and update beliefs for vertices of the current color.
+        val newBeliefCol = when(v("color") === color && aggregates("aggMess").isNotNull,
+          logistic(aggregates("aggMess") + v("a")))
+          .otherwise(v("belief"))  // keep old beliefs for other colors
+        val newVertices = v
```

Comment:
Is there any way we can do that for the user? The GraphX API returns a VertexRDD, but it seems more natural to me to return a new graph with updated vertices. This is essentially what you are doing here.